### ISSUE TITLE
Explanation of <T extends @Nullable Object> instead of <@Nullable T>

### ISSUE THREAD

**Issue by bjmi** [jspecify/jspecify#364]
Please spent a short paragraph in Nullness User Guide why I have to write
`<T extends @Nullable Object> T firstOrDefault(List<T> list, T defaultValue) {`
instead of
`<@Nullable T> T firstOrDefault(List<T> list, T defaultValue) {`  
and / or provide a comprehensible explanation what latter actually means.

(Out of scope: is it possible to pass `Consumer<@Nullable String> c` to `void process(Consumer<String> s)`)

**Comment by cpovirk** [comment]
You can find a little discussion of that [in the design FAQ](https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.e3zcvwu1vabk), which is currently linked from [the bottom of one page in the general FAQ in the wiki](https://github.com/jspecify/jspecify/wiki/jspecify-faq#i-have-more-questions-about-the-nullness-features). It's definitely going to need to be somewhere more prominent someday :)

For the `Consumer` question, see https://github.com/jspecify/jspecify/issues/72.

**Comment by kevinb9n** [comment]
Thanks for the feedback! I'd like to make sure we have enough "(Why?)" breadcrumbs in the various documents to point people to the FAQ. You can see a few of those in the javadoc. I'll make this issue about doing that in the user guide too.

**Comment by kevinb9n** [comment]
(For the record Chris :-) the Design FAQ is also linked from the start-here page but it's very easy for the eyes to pass over that part.)

As for a comprehensible explanation of where `<T extends @Nullable Object>` comes from, the "understanding the null-aware type system" doc will try to tackle that. Its early unsatisfactory form is [here](https://github.com/jspecify/jspecify/wiki/null-aware-type-system) .. I will work more on it, give it some diagrams and so forth.

**Comment by patpatpat123** [comment]
# Proposal: "(Why?)" links for the User Guide

Following up on this issue, I've drafted a change to `docs/docs/user-guide.md` (the source of [jspecify.dev/docs/user-guide](https://jspecify.dev/docs/user-guide)) and would like feedback on the approach before opening a PR.

## Approach

Per @kevinb9n's reframing, the change adds "(Why?)" breadcrumbs to the user guide like the javadoc already has — with one improvement over the javadoc pattern: instead of a single generic link to the Design FAQ (the javadoc's `bit.ly/3ppb8ZC` lands at the top of the doc), **each link deep-links to the specific FAQ question that answers it**, using the same `#heading=` anchor style @cpovirk used in his comment above.

Six links are added, matching the surprising design decisions the guide walks through:

| User guide location | Links to |
|---|---|
| "unspecified nullness" introduction ("What about unannotated types?") | wiki: [nullness-unspecified](https://github.com/jspecify/jspecify/wiki/nullness-unspecified) |
| `@NullMarked` makes unannotated types non-null | FAQ: *"Why does being null-marked require us to explicitly mark our nullable types, rather than marking the non-null ones?"* |
| Local variable root types aren't annotated | FAQ: *"…but then that falls apart for local variables? What gives?"* |
| `<E extends @Nullable Object>` (Declaring generic types) | FAQ: *"Then why can't we at least write `class MyList<@Nullable T>`?"* |
| `ImmutableList<E>` bound defaults to non-null | FAQ: *"So `List<?>` might admit nullable elements, but a declaration like `class List<T>` doesn't?"* |
| `<?>` vs. `<? extends Object>` (Wildcard bounds) | FAQ: *"Why are you making `List<?>` and `List<? extends Object>` mean different things?"* |

## Design notes

*   **The original question in this issue gets a self-describing breadcrumb.** At the exact spot where the guide says "Often this will be `<E extends @Nullable Object>`", the link text is *"Why can't it just be `<@Nullable E>`?"* rather than a bare "(Why?)", since that's the precise question readers (like the OP) are asking at that point. The other five use the javadoc's plain "(Why?)" style.
*   **The "unspecified nullness" link points at the wiki article directly** rather than the FAQ entry, because that FAQ entry's answer is itself just a pointer to [wiki/nullness-unspecified](https://github.com/jspecify/jspecify/wiki/nullness-unspecified) — this saves readers a hop, and matches what the `@NullUnmarked` javadoc does.
*   **A short "Further reading" section is added at the end of the guide**, linking the full [Nullness Design FAQ](https://github.com/jspecify/jspecify/wiki/nullness-design-FAQ) and the [null-aware type system](https://github.com/jspecify/jspecify/wiki/null-aware-type-system) article. This addresses @cpovirk's "it's definitely going to need to be somewhere more prominent someday" and @kevinb9n's pointer to the null-aware-type-system doc for the deeper explanation of where `<T extends @Nullable Object>` comes from.
*   **All URLs are reference-style link definitions at the bottom of the file**, so if/when the Design FAQ migrates off Google Docs to a better home, there's a single place to update.

## Verification

*   All `#heading=h.…` anchors were checked against the current Design FAQ doc (they're stable per-heading IDs; the one @cpovirk posted in 2023 still resolves today).
*   `npm run build` (Docusaurus) passes, and all six links render correctly in the built HTML.

## Diff

````diff
diff --git a/docs/docs/user-guide.md b/docs/docs/user-guide.md
index 6797ba5..b49a531 100644
--- a/docs/docs/user-guide.md
+++ b/docs/docs/user-guide.md
@@ -141,6 +141,7 @@ A type like `String` that isn't annotated with either `@Nullable` or `@NonNull`
 means what it always used to mean: its values might be intended to include
 `null` or might not, depending on whatever documentation you can find (but see
 [below](#nullmarked) for help!). JSpecify calls this "unspecified nullness".
+([Why?][why-unspecified-nullness])
 
 ```java
 class Unannotated {
@@ -160,7 +161,8 @@ code with either `@Nullable` or `@NonNull` to avoid unspecified nullness
 
 So JSpecify gives you the [`@NullMarked`] annotation. When you apply
 `@NullMarked` to a module, package, class, or method, it means that unannotated
-types in that scope are treated as if they were annotated with `@NonNull`.
+types in that scope are treated as if they were annotated with `@NonNull`
+([Why?][why-non-null-default]).
 (Below we will see that there are some exceptions to this for
 [local variables](#local-variables) and [type variables](#declaring-generics).)
 In code covered by `@NullMarked`, `String x` means the same as `@NonNull String
@@ -217,7 +219,8 @@ within that narrower scope.
 `@Nullable` and `@NonNull` aren't applied to local variables—at least not their
 root types. (They should be applied to type arguments and array components.) The
 reason is that it is possible to *infer* whether a variable can be `null` based
-on the values that are assigned to the variable. For example:
+on the values that are assigned to the variable ([Why?][why-local-variables]).
+For example:
 
 ```java
 @NullMarked
@@ -309,9 +312,11 @@ would not be compatible with `<E extends Object>`.
 The implication of all this is that every time you define a type variable like
 `E` you need to decide whether it can be substituted with a `@Nullable` type. If
 it can, then it must have a `@Nullable` bound. Often this will be `<E extends
-@Nullable Object>`. On the other hand, if it *can't* represent a `@Nullable`
-type, that is expressed by not having `@Nullable` in its bound (including the
-case of not having an explicit bound at all). Here's another example:
+@Nullable Object>` ([Why can't it just be
+`<@Nullable E>`?][why-no-nullable-type-parameter]). On the other hand, if it
+*can't* represent a `@Nullable` type, that is expressed by not having
+`@Nullable` in its bound (including the case of not having an explicit bound at
+all). Here's another example:
 
 ```java
 @NullMarked
@@ -321,7 +326,7 @@ public class ImmutableList<E> implements List<E> {...}
 Here, because it is `ImmutableList<E>` and not `ImmutableList<E extends
 @Nullable Object>`, it is not legal to write `ImmutableList<@Nullable String>`.
 You can only write `ImmutableList<String>`, which is a list of non-null `String`
-references.
+references. ([Why?][why-bounds-non-null])
 
 ### Using type variables in generic types
 
@@ -487,7 +492,8 @@ extends @Nullable Number>` means a list where the elements can be `null`, and
 However, there's a difference when there is no explicit bound. We saw that a
 type variable definition like `<E>` means `<E extends Object>` and that means it
 is not `@Nullable`. But `<?>` actually means `<? extends B>`, where `B` is the
-bound of the corresponding type variable. So if we have
+bound of the corresponding type variable ([Why?][why-unbounded-wildcards]). So
+if we have
 
 ```java
 interface List<E extends @Nullable Object> {...}
@@ -509,7 +515,23 @@ so `ImmutableList<?>` means the same as `ImmutableList<? extends Object>`. And
 here, `@NullMarked` means that `Object` excludes `null`. The `get(int)` method
 of `List<?>` can return `null` but the same method of `ImmutableList<?>` can't.
 
+## Further reading
+
+The "(Why?)" links in this guide point to individual entries in the
+[Nullness Design FAQ], which answers many more questions about the reasoning
+behind JSpecify's design decisions. And for a deeper explanation of how these
+annotations effectively make Java's type system null-aware, see the
+[null-aware type system] article in our wiki.
+
 [`@NonNull`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NonNull.html
 [`@Nullable`]: https://jspecify.dev/docs/api/org/jspecify/annotations/Nullable.html
 [`@NullMarked`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NullMarked.html
 [`@NullUnmarked`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NullUnmarked.html
+[null-aware type system]: https://github.com/jspecify/jspecify/wiki/null-aware-type-system
+[nullness design faq]: https://github.com/jspecify/jspecify/wiki/nullness-design-FAQ
+[why-bounds-non-null]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.15oks82r02g5
+[why-local-variables]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.lvo4aaw4bkyk
+[why-no-nullable-type-parameter]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.e3zcvwu1vabk
+[why-non-null-default]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.elscixoom7a5
+[why-unbounded-wildcards]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.s0il13hj8ves
+[why-unspecified-nullness]: https://github.com/jspecify/jspecify/wiki/nullness-unspecified
````

Happy to open this as a PR if the approach looks reasonable — or adjust the link targets/wording per your preferences.

**Comment by cpovirk** [comment]
Thank you, that looks very helpful!

I have only one thought about your choice of links: For why-no-nullable-type-parameter, I might link h.15oks82r02g5 (as you do for why-bounds-non-null; maybe they could share a reference, or maybe it would make sense to remove one of the two two "Why? links) instead of the current link (which is also relevant but more about the verbosity of the alternative than the choice itself, I'd say).

A PR is great. There is https://cla.developers.google.com/ for you to navigate before we can actually merge anything, but if you're able to sign that, I'd be happy to merge a PR.

**Comment by patpatpat123** [comment]
Thanks @cpovirk ,

I will not be able to sign. But it is ok, my goal is just to help the project.

You're right about that link. Re-reading the two FAQ entries, `h.e3zcvwu1vabk` really answers "why is the required spelling so verbose?", while the decision the guide is explaining at that point (bounds are non-null by default, so a nullable bound has to be spelled out) is `h.15oks82r02g5`. I've updated the draft accordingly.

## What changed since the last diff

*   The link at "Often this will be `<E extends @Nullable Object>`" now points to `h.15oks82r02g5` ("So `List<?>` might admit nullable elements, but a declaration like `class List<T>` doesn't?"), sharing the `[why-bounds-non-null]` reference as you suggested. I also went back to a plain "(Why?)" for the link text, since the "Why can't it just be `<@Nullable E>`?" wording no longer matched the target.
*   With both links then pointing at the same entry a few lines apart, I took your other suggestion too and dropped the second one (after the `ImmutableList<String>` example). So the guide now has five "(Why?)" links instead of six.
*   Removed the now-unused reference definition.

One nice side effect: readers who *are* wondering about the `<@Nullable E>` shorthand still land in the right neighborhood, since that entry comes immediately after `h.15oks82r02g5` in the FAQ doc.

Re-verified: the Docusaurus build passes and all five links render correctly in the built HTML.

On the PR/CLA: I need to check what signing the Google CLA looks like from my side (corporate email, so there may be some process to navigate). In the meantime the full revised diff is below — and if it's simpler for you to just pick it up and apply it directly, please feel absolutely free!

## Revised diff

````diff
diff --git a/docs/docs/user-guide.md b/docs/docs/user-guide.md
index 6797ba5..4193d22 100644
--- a/docs/docs/user-guide.md
+++ b/docs/docs/user-guide.md
@@ -141,6 +141,7 @@ A type like `String` that isn't annotated with either `@Nullable` or `@NonNull`
 means what it always used to mean: its values might be intended to include
 `null` or might not, depending on whatever documentation you can find (but see
 [below](#nullmarked) for help!). JSpecify calls this "unspecified nullness".
+([Why?][why-unspecified-nullness])
 
 ```java
 class Unannotated {
@@ -160,7 +161,8 @@ code with either `@Nullable` or `@NonNull` to avoid unspecified nullness
 
 So JSpecify gives you the [`@NullMarked`] annotation. When you apply
 `@NullMarked` to a module, package, class, or method, it means that unannotated
-types in that scope are treated as if they were annotated with `@NonNull`.
+types in that scope are treated as if they were annotated with `@NonNull`
+([Why?][why-non-null-default]).
 (Below we will see that there are some exceptions to this for
 [local variables](#local-variables) and [type variables](#declaring-generics).)
 In code covered by `@NullMarked`, `String x` means the same as `@NonNull String
@@ -217,7 +219,8 @@ within that narrower scope.
 `@Nullable` and `@NonNull` aren't applied to local variables—at least not their
 root types. (They should be applied to type arguments and array components.) The
 reason is that it is possible to *infer* whether a variable can be `null` based
-on the values that are assigned to the variable. For example:
+on the values that are assigned to the variable ([Why?][why-local-variables]).
+For example:
 
 ```java
 @NullMarked
@@ -309,9 +312,10 @@ would not be compatible with `<E extends Object>`.
 The implication of all this is that every time you define a type variable like
 `E` you need to decide whether it can be substituted with a `@Nullable` type. If
 it can, then it must have a `@Nullable` bound. Often this will be `<E extends
-@Nullable Object>`. On the other hand, if it *can't* represent a `@Nullable`
-type, that is expressed by not having `@Nullable` in its bound (including the
-case of not having an explicit bound at all). Here's another example:
+@Nullable Object>` ([Why?][why-bounds-non-null]). On the other hand, if it
+*can't* represent a `@Nullable` type, that is expressed by not having
+`@Nullable` in its bound (including the case of not having an explicit bound at
+all). Here's another example:
 
 ```java
 @NullMarked
@@ -487,7 +491,8 @@ extends @Nullable Number>` means a list where the elements can be `null`, and
 However, there's a difference when there is no explicit bound. We saw that a
 type variable definition like `<E>` means `<E extends Object>` and that means it
 is not `@Nullable`. But `<?>` actually means `<? extends B>`, where `B` is the
-bound of the corresponding type variable. So if we have
+bound of the corresponding type variable ([Why?][why-unbounded-wildcards]). So
+if we have
 
 ```java
 interface List<E extends @Nullable Object> {...}
@@ -509,7 +514,22 @@ so `ImmutableList<?>` means the same as `ImmutableList<? extends Object>`. And
 here, `@NullMarked` means that `Object` excludes `null`. The `get(int)` method
 of `List<?>` can return `null` but the same method of `ImmutableList<?>` can't.
 
+## Further reading
+
+The "(Why?)" links in this guide point to individual entries in the
+[Nullness Design FAQ], which answers many more questions about the reasoning
+behind JSpecify's design decisions. And for a deeper explanation of how these
+annotations effectively make Java's type system null-aware, see the
+[null-aware type system] article in our wiki.
+
 [`@NonNull`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NonNull.html
 [`@Nullable`]: https://jspecify.dev/docs/api/org/jspecify/annotations/Nullable.html
 [`@NullMarked`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NullMarked.html
 [`@NullUnmarked`]: https://jspecify.dev/docs/api/org/jspecify/annotations/NullUnmarked.html
+[null-aware type system]: https://github.com/jspecify/jspecify/wiki/null-aware-type-system
+[nullness design faq]: https://github.com/jspecify/jspecify/wiki/nullness-design-FAQ
+[why-bounds-non-null]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.15oks82r02g5
+[why-local-variables]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.lvo4aaw4bkyk
+[why-non-null-default]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.elscixoom7a5
+[why-unbounded-wildcards]: https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview#heading=h.s0il13hj8ves
+[why-unspecified-nullness]: https://github.com/jspecify/jspecify/wiki/nullness-unspecified
````


Would you be able to make the PR instead?

**Comment by kevinb9n** [comment]
The deep links are a good thing!

It also reminds me that we should probably just migrate the Google Doc to a real markdown page -- I suspect that's the only reason the javadoc didn't bother with the deep links yet, because I thought the links would be changing.

**Comment by cpovirk** [comment]
@kevinb9n, is https://jspecify.dev/docs/nullness-design-faq/ a good URL in your book? If so, I can send you a PR, and we can iterate on section titles and other minutiae?

(previously on Javadoc deep links: https://github.com/jspecify/jspecify/issues/408)

\[edit: Oops, I very belatedly see this: "In the unspecified future this will become markdown hosted at `jspecify.org`, and as we update links we will make them into deep-links to the appropriate section (sorry that for now you were brought only to the top).")

@patpatpat123, I should probably have avoided reviewing the diffs entirely without a signed CLA: I am not a lawyer, but I suspect that the policy is closer to "We can't apply specific changes suggested without a signed CLA" than "We can't accept things in GitHub-PR form without a signed CLA." If you're able to ultimately get it signed, then that will be easiest, but maybe someone else will attempt some similar changes in the future once we have a Markdown doc with a better link format.

**Comment by patpatpat123** [comment]
Well, whoever wants to use what I wrote for a PR, please feel free to do so. The goal is to improve this project
